File: algorithm-dynamic-programming-implementation-enhanced/src/algorithms/knapsack01.ts
```typescript
/**
 * @fileoverview Implementations for the 0/1 Knapsack Problem.
 * Given n items, each with a weight[i] and a value[i], and a knapsack with a maximum capacity,
 * determine the maximum total value that can be carried in the knapsack such that
 * the total weight does not exceed the capacity. Each item can either be included or not included.
 */

/**
 * 1. Recursive (Brute Force)
 * Solves the 0/1 Knapsack problem using a naive recursive approach.
 * For each item, it has two choices: include it or exclude it.
 * This leads to recomputing solutions for the same subproblems.
 *
 * Time Complexity: O(2^n) - Exponential, as for each item, we explore two branches.
 * Space Complexity: O(n) - Due to recursion stack depth.
 *
 * @param weights An array of weights for each item.
 * @param values An array of values for each item.
 * @param capacity The maximum capacity of the knapsack.
 * @param n The current number of items to consider (initially total items).
 * @returns The maximum value that can be carried.
 */
export function knapsack01_recursive(
    weights: number[],
    values: number[],
    capacity: number,
    n: number = weights.length
): number {
    // Base case: If no items left or knapsack capacity is 0, no value can be obtained.
    if (n === 0 || capacity === 0) {
        return 0;
    }

    // If the weight of the current item (n-1) is more than the knapsack capacity,
    // this item cannot be included. Move to the next item.
    if (weights[n - 1] > capacity) {
        return knapsack01_recursive(weights, values, capacity, n - 1);
    } else {
        // Two choices for the current item (n-1):
        // 1. Include the item:
        //    Add its value and recursively call for remaining capacity and items.
        const includeItem =
            values[n - 1] + knapsack01_recursive(weights, values, capacity - weights[n - 1], n - 1);

        // 2. Exclude the item:
        //    Recursively call for the same capacity and remaining items.
        const excludeItem = knapsack01_recursive(weights, values, capacity, n - 1);

        // Return the maximum of the two choices.
        return Math.max(includeItem, excludeItem);
    }
}

/**
 * 2. Memoization (Top-Down Dynamic Programming)
 * Solves the 0/1 Knapsack problem using memoization to store results of subproblems.
 * `memo[n][w]` stores the maximum value for `n` items and `w` capacity.
 *
 * Time Complexity: O(n * capacity) - Each state `(n, w)` is computed once.
 * Space Complexity: O(n * capacity) - For the memoization table and recursion stack.
 *
 * @param weights An array of weights for each item.
 * @param values An array of values for each item.
 * @param capacity The maximum capacity of the knapsack.
 * @returns The maximum value that can be carried.
 */
export function knapsack01_memoized(
    weights: number[],
    values: number[],
    capacity: number
): number {
    const n = weights.length;
    // Initialize memoization table with -1 to indicate uncomputed states.
    // memo[i][j] will store the max value using first 'i' items with 'j' capacity.
    const memo: number[][] = Array(n + 1)
        .fill(0)
        .map(() => Array(capacity + 1).fill(-1));

    function solve(currentItemIndex: number, currentCapacity: number): number {
        // Base case: If no items left or knapsack capacity is 0, no value.
        if (currentItemIndex === 0 || currentCapacity === 0) {
            return 0;
        }

        // Check if the result is already computed
        if (memo[currentItemIndex][currentCapacity] !== -1) {
            return memo[currentItemIndex][currentCapacity];
        }

        // If current item's weight is more than current capacity, skip it.
        if (weights[currentItemIndex - 1] > currentCapacity) {
            memo[currentItemIndex][currentCapacity] = solve(currentItemIndex - 1, currentCapacity);
            return memo[currentItemIndex][currentCapacity];
        } else {
            // Two choices: include or exclude the current item
            const includeItem =
                values[currentItemIndex - 1] +
                solve(currentItemIndex - 1, currentCapacity - weights[currentItemIndex - 1]);
            const excludeItem = solve(currentItemIndex - 1, currentCapacity);

            memo[currentItemIndex][currentCapacity] = Math.max(includeItem, excludeItem);
            return memo[currentItemIndex][currentCapacity];
        }
    }

    // Start with all items and full capacity
    return solve(n, capacity);
}

/**
 * 3. Tabulation (Bottom-Up Dynamic Programming)
 * Solves the 0/1 Knapsack problem iteratively using a 2D DP table.
 * `dp[i][j]` represents the maximum value that can be obtained using the first `i` items
 * with a knapsack capacity of `j`.
 *
 * `dp[i][j] = max(dp[i-1][j], values[i-1] + dp[i-1][j - weights[i-1]])`
 *
 * Time Complexity: O(n * capacity) - Two nested loops.
 * Space Complexity: O(n * capacity) - For the DP table.
 *
 * @param weights An array of weights for each item.
 * @param values An array of values for each item.
 * @param capacity The maximum capacity of the knapsack.
 * @returns The maximum value that can be carried.
 */
export function knapsack01_tabulated(
    weights: number[],
    values: number[],
    capacity: number
): number {
    const n = weights.length;
    // dp[i][j] will store the maximum value for first 'i' items with capacity 'j'.
    // The DP table has (n+1) rows and (capacity+1) columns.
    const dp: number[][] = Array(n + 1)
        .fill(0)
        .map(() => Array(capacity + 1).fill(0));

    // Fill the DP table
    // Iterate over items (from 1 to n)
    for (let i = 1; i <= n; i++) {
        // Iterate over capacities (from 1 to capacity)
        for (let j = 1; j <= capacity; j++) {
            const currentItemWeight = weights[i - 1]; // Current item's weight
            const currentItemValue = values[i - 1];   // Current item's value

            // If the current item's weight is less than or equal to the current capacity 'j'
            if (currentItemWeight <= j) {
                // We have two options:
                // 1. Include the current item: value + max value from previous items with remaining capacity.
                const valueWithItem = currentItemValue + dp[i - 1][j - currentItemWeight];
                // 2. Exclude the current item: value from previous items with the same capacity.
                const valueWithoutItem = dp[i - 1][j];
                dp[i][j] = Math.max(valueWithItem, valueWithoutItem);
            } else {
                // If the current item's weight is greater than the current capacity 'j',
                // we cannot include this item. So, we take the value from previous items.
                dp[i][j] = dp[i - 1][j];
            }
        }
    }

    // The result is in the bottom-right cell of the DP table.
    return dp[n][capacity];
}

/**
 * 4. Space-Optimized Tabulation
 * Solves the 0/1 Knapsack problem with reduced space complexity.
 * Since `dp[i][j]` only depends on `dp[i-1][...]` (values from the previous row),
 * we can optimize space to O(capacity) by only maintaining two rows (current and previous)
 * or even a single row if we iterate carefully from right to left for capacity.
 *
 * Time Complexity: O(n * capacity) - Two nested loops.
 * Space Complexity: O(capacity) - Uses a single 1D array of size `capacity + 1`.
 *
 * @param weights An array of weights for each item.
 * @param values An array of values for each item.
 * @param capacity The maximum capacity of the knapsack.
 * @returns The maximum value that can be carried.
 */
export function knapsack01_spaceOptimized(
    weights: number[],
    values: number[],
    capacity: number
): number {
    const n = weights.length;
    // dp array represents the current row's maximum values for each capacity.
    // dp[j] will store the maximum value for current items with capacity 'j'.
    const dp: number[] = new Array(capacity + 1).fill(0);

    // Iterate over items
    for (let i = 0; i < n; i++) {
        const currentItemWeight = weights[i];
        const currentItemValue = values[i];

        // Iterate over capacity from right to left.
        // This is crucial for space optimization. If we iterate left to right,
        // dp[j - currentItemWeight] would already contain the value for the *current* item,
        // effectively making it an unbounded knapsack. By going right to left,
        // dp[j - currentItemWeight] still holds the value from the *previous* item's consideration,
        // ensuring each item is considered only once (0/1 property).
        for (let j = capacity; j >= currentItemWeight; j--) {
            // max(value without current item, value with current item)
            // dp[j] (without current item) is its value from the previous iteration (i-1)
            // dp[j - currentItemWeight] (from previous iteration i-1) + currentItemValue
            dp[j] = Math.max(dp[j], dp[j - currentItemWeight] + currentItemValue);
        }
    }

    // The result is the maximum value for the full capacity after considering all items.
    return dp[capacity];
}
```

File: coding-interview-dynamic-programming-advanced/dp_interview_project/algorithms/knapsack_01.py
```python
import functools
from typing import List, Tuple

class Knapsack01:
    """
    Implementations for the 0/1 Knapsack Problem.
    Given weights and values of N items, put some items in a knapsack of capacity W
    to get the maximum total value in the knapsack. Each item can only be picked once.
    """

    @staticmethod
    def knapsack_recursive_bruteforce(weights: List[int], values: List[int], capacity: int) -> int:
        """
        Solves the 0/1 Knapsack problem using a naive recursive (brute-force) approach.
        At each item, we have two choices: either include it or exclude it.

        Args:
            weights (List[int]): List of weights of the items.
            values (List[int]): List of values of the items.
            capacity (int): The maximum capacity of the knapsack.

        Returns:
            int: The maximum total value that can be obtained.

        Time Complexity: O(2^N) - For N items, at each item we have 2 choices, leading to exponential complexity.
        Space Complexity: O(N) - Due to the recursion stack depth.
        """
        n = len(weights)

        def solve(idx, current_capacity):
            # Base Case: If no items left or capacity is 0
            if idx == n or current_capacity == 0:
                return 0

            # If the weight of the current item exceeds the current capacity, skip it
            if weights[idx] > current_capacity:
                return solve(idx + 1, current_capacity)
            else:
                # Option 1: Include the current item
                # Value of current item + max value from remaining items with reduced capacity
                include_item = values[idx] + solve(idx + 1, current_capacity - weights[idx])

                # Option 2: Exclude the current item
                # Max value from remaining items with same capacity
                exclude_item = solve(idx + 1, current_capacity)

                return max(include_item, exclude_item)

        return solve(0, capacity)

    @staticmethod
    @functools.lru_cache(maxsize=None)
    def knapsack_memoization(weights_tuple: Tuple[int, ...], values_tuple: Tuple[int, ...], capacity: int) -> int:
        """
        Solves the 0/1 Knapsack problem using memoization (top-down Dynamic Programming).
        Stores the results of subproblems (idx, current_capacity) to avoid redundant computations.

        Args:
            weights_tuple (Tuple[int, ...]): Tuple of weights (must be hashable for lru_cache).
            values_tuple (Tuple[int, ...]): Tuple of values (must be hashable for lru_cache).
            capacity (int): The maximum capacity of the knapsack.

        Returns:
            int: The maximum total value that can be obtained.

        Time Complexity: O(N * W) - Where N is the number of items and W is the knapsack capacity.
                                     Each state (idx, current_capacity) is computed once.
        Space Complexity: O(N * W) - For the memoization table (cache) and recursion stack.
        """
        weights = list(weights_tuple) # Convert back to list for indexing if needed, but not strictly
        values = list(values_tuple)   # necessary if we only use idx

        n = len(weights)

        # Inner recursive function to be memoized
        # We need to pass immutable states (idx, current_capacity) to lru_cache
        # weights and values are implicitly available in the closure
        def solve_memo(idx, current_cap):
            # Base Case: No items left or no capacity
            if idx == n or current_cap == 0:
                return 0

            # If the weight of the current item exceeds the current capacity, skip it
            if weights[idx] > current_cap:
                return solve_memo(idx + 1, current_cap)
            else:
                # Option 1: Include the current item
                include_item = values[idx] + solve_memo(idx + 1, current_cap - weights[idx])

                # Option 2: Exclude the current item
                exclude_item = solve_memo(idx + 1, current_cap)

                return max(include_item, exclude_item)

        # lru_cache works on the function itself, not directly on a method's inner function.
        # So we wrap the actual logic in an auxiliary function `_knapsack_memo_helper`
        # and then apply lru_cache to it.
        # Alternatively, the outer function could be directly decorated if it takes `idx` and `current_cap`
        # as arguments directly. For a class method, we need a slight adjustment.
        # For simplicity, we make the memoized function an internal helper.
        # The key for lru_cache needs to be the actual arguments changing, which are `idx` and `current_cap`.
        # So we define `_solve_memo` and decorate it inside the static method.

        # The trick for lru_cache with static methods and external context is to pass
        # the context (weights, values, n) implicitly or to use a closure carefully.
        # Here, the `knapsack_memoization` method itself is decorated and acts as the entry point.
        # We need to make sure the arguments `weights_tuple`, `values_tuple`, `capacity` are part of the cache key.
        # However, typically memoization is applied to the recursive helper function `solve_memo`.

        # Let's adjust for a more standard lru_cache application.
        # lru_cache cannot directly cache a method with `self` or `cls` arguments,
        # and it needs hashable arguments. The weights_tuple and values_tuple
        # are made hashable for the outer call, but the inner recursion depends on `idx` and `current_cap`.

        # To properly use lru_cache for the inner recursive state:
        @functools.lru_cache(maxsize=None)
        def _solve_memo_inner(idx, current_cap):
            if idx == n or current_cap == 0:
                return 0
            if weights[idx] > current_cap:
                return _solve_memo_inner(idx + 1, current_cap)
            else:
                include_item = values[idx] + _solve_memo_inner(idx + 1, current_cap - weights[idx])
                exclude_item = _solve_memo_inner(idx + 1, current_cap)
                return max(include_item, exclude_item)

        # We call the memoized inner function
        return _solve_memo_inner(0, capacity)


    @staticmethod
    def knapsack_tabulation(weights: List[int], values: List[int], capacity: int) -> int:
        """
        Solves the 0/1 Knapsack problem using tabulation (bottom-up Dynamic Programming).
        Builds a DP table where dp[i][w] represents the maximum value that can be
        obtained with the first 'i' items and a knapsack capacity of 'w'.

        Args:
            weights (List[int]): List of weights of the items.
            values (List[int]): List of values of the items.
            capacity (int): The maximum capacity of the knapsack.

        Returns:
            int: The maximum total value that can be obtained.

        Time Complexity: O(N * W) - Where N is the number of items and W is the knapsack capacity.
                                     Two nested loops iterate N*W times.
        Space Complexity: O(N * W) - For the 2D DP table. Can be optimized to O(W) if only the
                                      previous row is needed.
        """
        n = len(weights)

        # dp[i][w] will store the maximum value for first 'i' items with capacity 'w'
        # dp table size: (n+1) rows, (capacity+1) columns
        dp = [[0 for _ in range(capacity + 1)] for _ in range(n + 1)]

        # Fill the dp table
        # i iterates through items (from 1 to n, 0-indexed item is i-1)
        # w iterates through capacities (from 0 to capacity)
        for i in range(1, n + 1):
            for w in range(capacity + 1):
                # Current item's weight and value
                current_item_weight = weights[i - 1]
                current_item_value = values[i - 1]

                # Case 1: If current item's weight is more than current capacity 'w'
                # We cannot include this item, so take value from the previous item (dp[i-1][w])
                if current_item_weight > w:
                    dp[i][w] = dp[i - 1][w]
                # Case 2: We can include the current item
                # We have two choices:
                #   a) Exclude the current item: dp[i-1][w]
                #   b) Include the current item: current_item_value + dp[i-1][w - current_item_weight]
                #      (value of current item + max value of remaining capacity from previous items)
                else:
                    dp[i][w] = max(dp[i - 1][w], current_item_value + dp[i - 1][w - current_item_weight])

        return dp[n][capacity]

    @staticmethod
    def knapsack_tabulation_space_optimized(weights: List[int], values: List[int], capacity: int) -> int:
        """
        Solves the 0/1 Knapsack problem using space-optimized tabulation (bottom-up Dynamic Programming).
        Reduces the space complexity from O(N*W) to O(W) by only using one row of the DP table.

        Args:
            weights (List[int]): List of weights of the items.
            values (List[int]): List of values of the items.
            capacity (int): The maximum capacity of the knapsack.

        Returns:
            int: The maximum total value that can be obtained.

        Time Complexity: O(N * W) - Two nested loops iterate N*W times.
        Space Complexity: O(W) - Only a 1D DP array of size (capacity+1) is used.
        """
        n = len(weights)

        # dp[w] will store the maximum value for the current set of items with capacity 'w'
        # Initialize with 0s for a capacity up to `capacity`.
        dp = [0] * (capacity + 1)

        # Iterate through each item
        for i in range(n):
            current_item_weight = weights[i]
            current_item_value = values[i]

            # Iterate through capacity from right to left
            # This is crucial for 0/1 Knapsack space optimization.
            # If we iterate left to right, we might use the current item multiple times
            # (treating it like unbounded knapsack) because dp[w - current_item_weight]
            # would already contain the value considering the current item.
            # Right-to-left ensures that dp[w - current_item_weight] refers to the
            # state *before* considering the current item.
            for w in range(capacity, current_item_weight - 1, -1):
                # Two choices for dp[w]:
                # 1. Exclude current item: dp[w] (value from previous iteration, before considering current item)
                # 2. Include current item: current_item_value + dp[w - current_item_weight]
                dp[w] = max(dp[w], current_item_value + dp[w - current_item_weight])

        return dp[capacity]


# Example Usage:
if __name__ == '__main__':
    print("--- 0/1 Knapsack Problem ---")

    # Example 1
    weights1 = [1, 2, 3]
    values1 = [6, 10, 12]
    capacity1 = 5
    print(f"\nItems: weights={weights1}, values={values1}, capacity={capacity1}")
    print(f"Brute Force: {Knapsack01.knapsack_recursive_bruteforce(weights1, values1, capacity1)}")
    print(f"Memoization: {Knapsack01.knapsack_memoization(tuple(weights1), tuple(values1), capacity1)}")
    Knapsack01.knapsack_memoization.cache_clear() # Clear cache for next calls if needed
    print(f"Tabulation: {Knapsack01.knapsack_tabulation(weights1, values1, capacity1)}")
    print(f"Space Optimized Tabulation: {Knapsack01.knapsack_tabulation_space_optimized(weights1, values1, capacity1)}")
    # Expected: max(10+0, 12+6) = 22 (items 2 and 3 if weights are 2 and 3, values 10 and 12, capacity 5)
    # Correct for example: W=1,V=6; W=2,V=10; W=3,V=12. Cap=5.
    # items 1 and 3 (weights 1+3=4, values 6+12=18)
    # items 2 and ? (weights 2, values 10)
    # If using first two: 1+2=3 < 5, 6+10=16
    # If using items 2 and 3: 2+3=5, 10+12=22. This is correct.

    # Example 2
    weights2 = [10, 20, 30]
    values2 = [60, 100, 120]
    capacity2 = 50
    print(f"\nItems: weights={weights2}, values={values2}, capacity={capacity2}")
    print(f"Brute Force: {Knapsack01.knapsack_recursive_bruteforce(weights2, values2, capacity2)}")
    print(f"Memoization: {Knapsack01.knapsack_memoization(tuple(weights2), tuple(values2), capacity2)}")
    Knapsack01.knapsack_memoization.cache_clear()
    print(f"Tabulation: {Knapsack01.knapsack_tabulation(weights2, values2, capacity2)}")
    print(f"Space Optimized Tabulation: {Knapsack01.knapsack_tabulation_space_optimized(weights2, values2, capacity2)}")
    # Expected: items 10 and 20 (weights 10+20=30, values 60+100=160).
    # If 20+30=50, values 100+120=220. This is the max.

    # Edge cases
    print("\n--- Edge Cases ---")
    print(f"Empty items: {Knapsack01.knapsack_tabulation([], [], 10)}") # Expected: 0
    print(f"Zero capacity: {Knapsack01.knapsack_tabulation([1,2], [10,20], 0)}") # Expected: 0
    print(f"Smallest item too heavy: {Knapsack01.knapsack_tabulation([100], [500], 50)}") # Expected: 0
    print(f"Single item, fits: {Knapsack01.knapsack_tabulation([10], [100], 10)}") # Expected: 100
```